### PR TITLE
User information no longer necessary in `render_template` calls

### PR DESCRIPTION
**Pull request by dvberkel**
This pull-request encompasses the following improvements:

* It places the user on the Flask session.
* It automatically injects user information in the context for templates.
* It refactors `is_teacher` and `is_admin` to accept a user instead of a request.

Below you can find a detailed motivation for these changes.

This pull-request uses a [_context processor_](https://flask.palletsprojects.com/en/2.0.x/templating/#context-processors) to inject user information, such as `username`, `is_teacher` and `is_admin` into the context automatically. This relieves developers from passing along this information with each call to `render_template`.

The context processor reads the user information from the user on the [Flask session](https://flask.palletsprojects.com/en/2.0.x/api/#flask.session). The user is placed on the session when they login and is removed when they user logs out or when they are destroyed.

The `is_teacher` and `is_admin` function accepted a request. This request was used to retrieve the current user from the database. This is an example of a function with side-effects, which should be avoided. Therefore the functions are changed to accept a user instead.

This pull-request is related to issue #1179. It removes a few of the globals that hamper progress on that front. 

## How to test
Since this is a refactoring, the application should work as before.

### CODE REVIEW

File: website/auth.py [hunk 1]
```diff
@@ -113,6 +113,8 @@ def login():
         if not check_password(body['password'], user['password']):
             return 'invalid username/password', 403
 
+        session['user'] = user
```
Comment (rix0rrr) [inline website/auth.py:116]: This should become a slice of the data.

I would prefer to automatically invalidate `is_teacher` (based on time).

File: app.py [hunk 2]
```diff
@@ -231,6 +231,12 @@ def check_language():
 def before_request_begin_logging():
     querylog.begin_global_log_record(path=request.path, method=request.method)
 
+# A context processor injects variables in the context that are available to all templates.
+@app.context_processor
+def enrich_context_with_user_info():
+    user = session.get('user', {'username': '', 'email': ''})
```
Comment (rix0rrr) [inline app.py:237]: This object should be standardized somewhere, probably close to the `copy_data_from_database_user` function...

Comment (dvberkel) [reply app.py:237]: As discussed:

We will repurpose the `current_user` function to return session user.

File: website/auth.py (hunk 1 above)
Comment (dvberkel) [reply website/auth.py]: The following slice of the data is placed in the current session
* `username`
* `email`
* `is_teacher`

Comment (dvberkel) [reply website/auth.py]: I have implemented a time-to-live for the current user

File: website/auth.py [hunk 3]
```diff
@@ -35,26 +35,52 @@ def hash(password, salt):
...
+# The main purpose of the current user is to provide a convenient container for
+# * username
+# * email
+# * is_teacher
+#
+# Since the is_teacher can be changed during a session we also store a time-to-live. When retrieving the current user, we can check if we need to reload data from the database.
+#
+# The current user should be retrieved with `current_user` function since it will return a sane default.
+# You can remove the current user from the Flask session with the `forget_current_user`.
+def create_current_user(db_user):
```
Comment (rix0rrr) [inline website/auth.py:48]: I would expect a function called `create_` to return something, but it has a side effect. Let's use a different verb?

```suggestion
def remember_current_user(db_user):
```

File: website/auth.py [hunk 4]
```diff
@@ -35,26 +35,52 @@ def hash(password, salt):
...
+# If the current user is to old, as determined by the time-to-live, we repopulate from the database.
+def current_user():
+    now = times()
+    user = session.get('user', {'username': '', 'email': ''})
+    ttl = session.get('user-ttl', None)
+    if ttl == None or now >= ttl:
+        username = user['username']
+        if username:
+            db_user = DATABASE.user_by_username(username)
+            create_current_user(db_user)
```
Comment (rix0rrr) [inline website/auth.py:66]:
```suggestion
            remember_current_user(db_user)
```

File: website/auth.py [hunk 5]
```diff
@@ -113,6 +139,8 @@ def login():
         if not check_password(body['password'], user['password']):
             return 'invalid username/password', 403
 
+        create_current_user(user)
```
Comment (rix0rrr) [inline website/auth.py:142]:
```suggestion
        remember_current_user(user)
```

**Review by rix0rrr — CHANGES_REQUESTED** [review]

File: website/auth.py (hunk 3 above)
Comment (dvberkel) [reply website/auth.py:48]: It puts a tear in my eye :man_cook: :kiss:

**Review by rix0rrr — DISMISSED** [review]

**Review by rix0rrr — APPROVED** [review]

**Pull request merged**